mrcnn/parallel_model.py

```diff
@@ -70,13 +78,19 @@ def make_parallel(self, inner_model,gpu_count):
         for i in range(gpu_count):
             with tf.device('/gpu:%d' % i):
                 with tf.name_scope('tower_%d' % i):
-                    # Run a slice of inputs through this replica
-                    zipped_inputs = zip(inner_model.input_names,
-                                        inner_model.inputs)
-                    inputs = [
-                        KL.Lambda(lambda s: input_slices[name][i],
-                                  output_shape=lambda s: (None,) + s[1:])(tensor)
-                        for name, tensor in zipped_inputs]
+                    inputs = []
+                    # Retrieve a slice of the input.
+                    for x in inner_model.inputs:
+                        # In-place input splitting which is not only
+                        # 5% ~ 12% faster but also less GPU memory
+                        # duplication.
+                        with tf.device(x.device):
+                            input_shape = K.int_shape(x)[1:]
+                            slice_i = KL.Lambda(get_slice,
```

Comment:
Since `x` is an input to this Lambda, and the Lambda OPs are on one of the GPUs while `x` is on the CPU since it's the input to the base model, then this would require TF to move `x` to the GPU to slice it there and then discard all but the one slice that's needed on that GPU. So this seems slower to me because it'll have to move more data to the GPU. What am I missing?

Reply:
Yes, idea is to have the split done on CPU, data moved to each resp. GPU, processing on the resp. GPU and the CPU is in charge on concatenating and else. I actually slightly modified keras code (see https://github.com/keras-team/keras/blob/master/keras/utils/multi_gpu_utils.py), for which the comments for multi_gpu_model is useful

Reply:
What I was trying to say in my previous comment is that the splitting is already being done on the CPU. The new changes make it happen on the GPU instead, and that's slower. I could be missing something, but it seems to me that the new changes are slower than the current state of the code. I'll read the links you posted below (thanks!)